Migrate released interactivity blocks to `@wordpress/interactivity`

samueljseay opened the pull request:
### Changes proposed in this Pull Request:

This feature branch encompasses the minimal changes needed to migrate existing (released) blocks using the interactivity API to using `@wordpress/interactivity`. 

Changes:

Blocks affected:

`woocommerce/store-notices`
`woocommerce/product-button`
`woocommerce/product-collection`
`woocommerce/product-gallery`

And also the product filters blocks

* Migrate imports to `@wordpress/interactivity`
* Migrate directives from `wc` to `wp`
* Create a new script modules build for those blocks
* Implement basic support for interactivity powered notices within `woocommerce/store-notices` block
* Utilize iapi powered store notices within `woocommerce/product-button`
* Also brings in https://github.com/woocommerce/woocommerce/pull/54979 to sync iAPI store with cart store


Closes https://github.com/woocommerce/woocommerce/issues/54208
Closes https://github.com/woocommerce/woocommerce/issues/54501
Closes https://github.com/woocommerce/woocommerce/issues/54502


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Test the affected blocks especially in frontend to confirm that they function as before
2. Test the new client-side store notices
3. Test that mini cart stays up to date with cart changes

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Migrate iAPI blocks to `@wordpress/interactivity`

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

samueljseay committed:
Add client side notices support to the woocommerce/store-notices block (#54552)

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

luisherranz committed:
Add to Cart (Product button) block: Prepare for migration to `@wordpress/interactivity` (#54576)

* Basic implementation of cart and addToCart in iAPI store

* Refactor stores to move stuff to the `woocommerce` one

* Minor refactorings

* Try updating syncpack

* Move server-side add to cart store logic to a shared class

* Load public woo store asyncronously

* Add refreshCart action to update the state when the page is cached

* Solve two minor eslint issues

* Add missing refreshCart action and a couple of extra questions

* Use the error response to throw

* Remove woocommerce_add_to_cart_quantity question

* Improve error handling

* Refactor and solve more questions

* Introduce retry with increasing timeout for refreshCartItems

* Remove unused loading styles

* Revert the use of inferred types

* Export Store type

* Fix merge conflict mistake.

* Fix another conflict mistake

* Fix more conflicts

* Remove hasCartLoaded and add displayViewCart back

* Move store notice logic to the new store part

---------

Co-authored-by: Sam Seay <samueljseay@gmail.com>

luisherranz committed:
Initial sync event implementation

luisherranz committed:
Move the sync events to the actions

samueljseay committed:
Move to script modules and @wordpress/interactivity for store-notices and product-button blocks (#54995)

samueljseay committed:
Update the lock file

samueljseay committed:
Move Product Gallery blocks to @wordpress/interactivity

samueljseay committed:
Enqueue the script module for large image

samueljseay committed:
Fix typos in paths

samueljseay committed:
Fix product gallery to work with wp interactivity

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Fix broken blocks

samueljseay committed:
Fix remaining interactivity directives to use wp prefix.

samueljseay committed:
Fix css lint issues

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Fix bugs in block hooking.

samueljseay committed:
Fix block hooking, dont duplicate, support GB or not.

samueljseay committed:
Fix up entries we still need to build store notices editor side

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Remove todos and mistake in webpack config

samueljseay committed:
Fix lints and update test selectors for wp prefix

samueljseay committed:
Fix lint issues

samueljseay committed:
Dont use console logging, fix some small typos.

samueljseay committed:
Merge branch 'dev/sync-wp-data-and-iapi-stores-54501', remote-tracking branch 'origin' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Update lock file

samueljseay committed:
Broaden response check to support new action that uses cart/items endpoint

samueljseay committed:
Trigger legacy added to cart event

samueljseay committed:
Fix imports

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Disable block script in product collection

kmanijak committed:
Lift context from inner block to Product Gallery to fix iAPI update (#55183)

Lift the context up

kmanijak committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductButton.php:
```diff
@@ -74,6 +70,7 @@ public function dequeue_add_to_cart_scripts() {
 	 * @return string Rendered block type output.
 	 */
 	protected function render( $attributes, $content, $block ) {
+		wp_enqueue_script_module( 'woocommerce/product-button' );
```
I better move this further down past the early return

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Skip experimental block tests that we don't explicitly need to pass right now

kmanijak committed:
Enqueue script module in large image

tjcafferkey committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

tjcafferkey committed:
Replace wc with wp in ProductGalleryLargeImageNextPrevious.php as it was lost in conflict resolution

luisherranz committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

luisherranz committed:
Add TS error exception and info

luisherranz commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductButton.php:
```diff
@@ -178,31 +177,22 @@ protected function render( $attributes, $content, $block ) {
-			data-wc-on--click="woocommerce/product-collection::actions.viewProduct"
-		';
+		$button_directives = 'data-wp-on--click="actions.addCartItem"';
+		$anchor_directive  = 'data-wp-on--click="woocommerce/product-collection::actions.viewProduct"';
```
This directive is also injected [from here](https://github.com/woocommerce/woocommerce/blob/04c998dc520c8112288c4babb67a6adfb88146bc/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/Controller.php#L155), so maybe we are repeating it?

Aljullu committed:
Add some improvements to the Add to Cart Button migration to @wordpress/interactivity (#55590)

* Update iAPI event names

* Remove unnecessary type cast

* Make refreshCartItems() async

* Add changelog file

* Revert "Make refreshCartItems() async"

This reverts commit 9312c42cb58881aaf9da431c4176106a8c1f1e4e.

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Update lockfile

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

luisherranz committed:
iAPI migration: Prevent quantity change notice showing on mini cart when adding items with the new iAPI sync event (#55295)

* Do not notify quantity changes when syncing

* Pass the quantityChanges data from the event

* Refactor

* Delete unnecessary code

* Avoid sync on getCartData resolver (again)

* Update thunk description

* Go back to modifying receiveCart

* Use quantityChanges if it exists

* Try using setTriggerStoreSyncEvent in the resolver

* Fix casing

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update pnpm-lock.yaml

* Wait until the mini cart is closed before adding next item to cart in e2e test.

* Update event names to match the scheme introduced in the base PR

* Change the page response wait on e2e test to match the new fetch to items endpoint

* Update all waits on the e2e dialog to be consistent

* Update lockfile

* Regenerate lockfile

* Change response matcher in e2e test

* Reorder the wait to wait for the request after adding cart item.

* Rearrange response waits and narrow down request method to reduce chance of racyness

---------

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Sam Seay <samueljseay@gmail.com>

kmanijak committed:
Migrate Product Collection from WC to WP iAPI (#55330)

* Fix iAPI navigation in Product Collection

* Extract dataset keys to a variavle

* Remove the loading bar animation and rely on iAPI one

* Fix lint

* Refactor the dataset access

* Fix lint

* Partially fix E2E test

* Move variables to if statements where they are used and fix types

* Remove code with accessibility messgaes as it's handled natively by iAPI

* Update comment from navigation-id to router-region

* Replace scroll calculations and functionality with a simple focus()

* Rely on core/router clientNavigationDisabled instead of custom attribute

* Refactor link validation and focus handling in product collection navigation

* Avoid using random values for data-wp-key

* Consolidate  an unnecessary function and fix PHP lint issue

* Don't parse/create unnecessary context

* Minor lint issue

* Don't overwrite context, as it is already been declared

* Consolidate all the directives in a single function

With so many functions, is difficult to know which directives are
being added and where. Also, it's unnecessary (and wasteful) to
initialize more than one HTML Tag Processor to add these directives.

---------

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>
Co-authored-by: Luis Herranz <luisherranz@gmail.com>

roykho committed:
[Experimental] Product filters: iapi migration (#55462)

* Change to use data-wp directive namespace
* Replace @woocommerce/interactivity with @wordpress/interactivity
* Add product filters block entries to webpack
* Enqueue module scripts
* Override block script and return null

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Enqueue product button script as late as possible.

samueljseay committed:
Unskip tests for filter blocks that have now been migrated to iAPI

gigitux commented:
Here are the results from my recent tests:

I created two fresh sites:
- Site 1: Runs the latest stable WooCommerce version.
- Site 2: Uses a WooCommerce build from this branch.

### ⚠️ Product Button

It looks like that there is a regression regarding the Product Button. On Site 2 for a second, I see only a dark button without any text inside. This doesn't happen on Site 1.


| Stable | Interactivity API branch |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/d775a60a-5d48-48b7-8328-57f539918495"></video>| <video src="https://github.com/user-attachments/assets/8b84874f-2c5a-4218-9561-03349be3aa1b"> </video>| 


### 👍 Product Collection

Pagination works as expected.

### 👍  Product Gallery

Everything works as expected.


### Store notices

I think that there isn't any to test yet mostly because we decide to pivot to [this approach](https://github.com/woocommerce/woocommerce/pull/55717), right?

samueljseay committed:
Ensure that product filters are skipped from building JS entry files in the old JS frontend build

samueljseay commented:
> I think that there isn't any to test yet mostly because we decide to pivot to https://github.com/woocommerce/woocommerce/pull/55717, right?

Yep that's right.

Will you look at the filters? I noticed some behaviour that looked odd, but Roy is not around to confirm with for another couple days.

> It looks like that there is a regression regarding the Product Button. On Site 2 for a second, I see only a dark button without any text inside. This doesn't happen on Site 1.

@gigitux Is this a local test or JN or other? Are you sure you rebuilt properly? It looks like the item is not being added to cart as well as the flash. No animation etc. Which is not something I can repro on the branch right now.

gigitux commented:
> Will you look at the filters? I noticed some behaviour that looked odd, but Roy is not around to confirm with for another couple days.
> 
> >

I can take a look at them if needed.

>@gigitux Is this a local test or JN or other? Are you sure you rebuilt properly? It looks like the item is not being added to cart as well as the flash. No animation etc. Which is not something I can repro on the branch right now. Can you also check console errors for me? Looks like something very wrong :D I'm sure you know this, but, running a watch build while swapping branches won't work, you'll need to restart the build.

I built a dedicated zip and uploaded it on a JN site, so I don't think that it is a build issue. I tried again with another JN site and I'm able to reproduce it. Can you reproduce it visiting https://totally-maximum-bee.jurassic.ninja/shop/ ?

luisherranz commented:
>  Can you reproduce it visiting https://totally-maximum-bee.jurassic.ninja/shop/ ?

That is working well on my computer, both in Chrome and Safari.

samueljseay committed:
Lock the woocommerce store, explicitly unlock when using it internally.

gigitux commented:
> > Can you reproduce it visiting https://totally-maximum-bee.jurassic.ninja/shop/ ?
> 
> That is working well on my computer, both in Chrome and Safari.

Don't you see any difference with trunk? On trunk, the text is immediately present, while on this branch for a second I see the button without text. I tested with Edge and I'm able to see this difference:

- https://real-weaver.jurassic.ninja/shop/ -> WooCommerce trunk branch
- https://totally-maximum-bee.jurassic.ninja/shop/ -> WooCommerce dev/migrate-to-wp-interactivity-54259 branch


| Trunk | Interactivity API  |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/81616e42-3870-4770-99ba-0eda1a8db522" /> | <video src="https://github.com/user-attachments/assets/c23585e3-4535-4958-baa0-e64ce133e04d" /> |

samueljseay commented:
> Can you reproduce it visiting https://totally-maximum-bee.jurassic.ninja/shop/ ?

@gigitux Works for me too.  I tested Chrome and Firefox too just in case.

samueljseay commented:
@gigitux oh I see what you mean yup. Your initial screen recording looked really broken so it was a distraction. Yes I see the brief flash with nothing in the button

gigitux commented:
> @gigitux oh I see what you mean yup. Your initial screen recording looked really broken so it was a distraction. Yes I see the brief flash with nothing in the button

Sorry for the confusion! I'm glad that you're able to replicate it! 


Also, I noticed that a spinner appears on the trunk when the user clicks the button, but it is no longer visible on this branch: is this a desired change?

samueljseay commented:
@gigitux that's a good question regarding the spinner. I'll find out. I like the new animation better but I don't know if we got sign off to change that.

@luisherranz regarding the issue with the text being empty, it seems like it should be initially filled by `$initial_product_text` but it seems like it gets wiped out before render. Any chance you could take a look at that? Even if I try render some static text in the span it is gone at render.

gigitux commented:
> @gigitux that's a good question regarding the spinner. I'll find out. I like the new animation better but I don't know if we got sign off to change that.
> 
> EDIT: The reason it's changed is because we optimistically update how many items are in the cart and that was a way to animate reverting the state if it failed. It would be weird to flash back to the old number. I think I probably need to ask more broadly about the animation change though.

I'm fine with this change. Thanks for the clarification!

Regarding the filters block, in some cases (reproducible by changing the max price via the input) the text is duplicated, as shown in the video below:  

https://github.com/user-attachments/assets/8f28ebb7-b5a0-4f57-bf86-7f0e84385556

If correct, @luisherranz is already aware of this bug. However, since the Filters Block is still experimental, I think that this shouldn't block the current PR.

luisherranz committed:
Reorder new syncpack packages

luisherranz committed:
Product Button: Fix missing derived state in the server

luisherranz commented:
> it seems like it should be initially filled by $initial_product_text but it seems like it gets wiped out before render. Any chance you could take a look at that?

Fixed in https://github.com/woocommerce/woocommerce/pull/55080/commits/68d4c16ebeb7aa0899bd7428775c588a3f3318c4.

samueljseay commented:
@luisherranz that's great, thank you. I'm curious why does server rendered text get wiped out before client side state is rendered? I expected it would work to have some server rendered text in there first.

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

luisherranz commented:
When the `data-wp-text` directive is evaluated on the server, it looks at the value that exists in `state.addToCardText` and updates the HTML with the value it finds. If that value is falsy, then it doesn't put any text.

The complexity here comes from the fact that the same HTML is being used for buttons that end up with `data-wp-text` and buttons that end up needing manual SSR, but when using the interactivity API, it's not necessary to do the SSR manually. You need to populate the state and derived state correctly, though.

More about Server Side Rendering in the Interactivity API in [this guide](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/server-side-rendering/).

samueljseay committed:
Move store initialization inside product button. (#55833)

samueljseay commented:
@luisherranz Thanks that makes sense

samueljseay committed:
Remove accidentally left in php imoport

samueljseay committed:
Add back skip to e2e test for unreleased filter blocks till it can be fixed.

luisherranz committed:
Add missing `supports.interactivity`

roykho commented:
Two issues I am noticing after the migration.

- In a mobile device (not dev tools), I can no longer tap "Filter products" to bring up the filters overlay.
- Once in the overlay, after selecting a filter such as the attribute filter "gray", the overlay closes itself instead of waiting for user to click/tap on "Apply" button. This is also what's causing the e2e test to fail.

I will take a look at it tomorrow but if you know/have any insights that would be great.

samueljseay committed:
Move iapi powered notices into product collection (#55717)

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Remove changelog that is not relevant

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay committed:
Merge branch 'trunk' into dev/migrate-to-wp-interactivity-54259

samueljseay commented:
> In a mobile device (not dev tools), I can no longer tap "Filter products" to bring up the filters overlay.
@roykho I cannot reproduce this. On an Android device running Chromium browser opening and closing the overlay works perfectly for me.

> Once in the overlay, after selecting a filter such as the attribute filter "gray", the overlay closes itself instead of waiting for user to click/tap on "Apply" button. This is also what's causing the e2e test to fail.
This one I can reproduce

samueljseay commented:
@roykho I just wanted to note that if you follow: p1740583588057979/1740556868.896719-slack-C05GY7MD3ST and disable the router region in ProductFilters, the overlay does not auto close, and also does not cause the duplication.

So it looks like it would be related to the bug mentioned there.

@luisherranz do you have a suggestion of a path forward here?

samueljseay commented:
@roykho hmm interesting. I have only tested the filters nested inside product collection so maybe that's the difference?

samueljseay commented:
@roykho Chrome on iOS is actually just webkit webview so this could be a safari problem? 🤔  i will see if i can test with safari

samueljseay commented:
I am EOD but tomorrow I'm thinking we'll be close to merge this, I just need to resolve conflicts and add a couple more changelog items. 

I want to see if I can repro the (possibly just safari) issue. We should maybe track it in an issue @roykho ?

roykho commented:
> In a mobile device (not dev tools), I can no longer tap "Filter products" to bring up the filters overlay.

@samueljseay I tested again back to back (trunk/this branch), I can't replicate it anymore. Strange!

So I think we're good to go.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/migrate-to-wp-interactivity-54259

luisherranz reviewed (approved):
Awesome work, folks 🙂👏
